src/Language/PureScript/Ide.hs

```diff
@@ -116,130 +112,84 @@ findPursuitPackages :: (MonadIO m) =>
-    traverse_ loadModule reexports
-    exportDepsModules <- catMaybes <$> traverse getModule reexports
-    exportDepDeps <- traverse loadReexports exportDepsModules
-    return $ concat exportDepDeps
-
-getDependenciesForModule :: Module -> [ModuleIdent]
-getDependenciesForModule (_, decls) = mapMaybe getDependencyName decls
-  where getDependencyName (Dependency dependencyName _ _) = Just dependencyName
-        getDependencyName _ = Nothing
-
-loadModule :: (PscIde m, MonadLogger m, MonadError PscIdeError m) =>
-              ModuleIdent -> m Text
-loadModule "Prim" = pure "Prim won't be loaded"
-loadModule mn = do
-  path <- filePathFromModule mn
-  loadExtern path
-  $(logDebug) ("Loaded extern file at: " <> T.pack path)
-  pure ("Loaded extern file at: " <> T.pack path)
-
-loadAllModules :: (PscIde m, MonadLogger m, MonadError PscIdeError m) => m Success
+-- | Takes the output directory and a filepath like "Monad.Control.Eff" and
+-- looks up, whether that folder contains an externs.json
+checkExternsPath :: FilePath -> FilePath -> IO (Maybe FilePath)
+checkExternsPath oDir d
+  | d `elem` [".", ".."] = pure Nothing
+  | otherwise = do
+      let file = oDir </> d </> "externs.json"
+      ex <- doesFileExist file
+      if ex
+        then pure (Just file)
+        else pure Nothing
+
+findAllExterns :: (Ide m, MonadError PscIdeError m) => m [FilePath]
+findAllExterns = do
+  oDir <- outputDirectory
+  unlessM (liftIO (doesDirectoryExist oDir))
+    (throwError (GeneralError "Couldn't locate your output directory."))
+  liftIO $ do
+    dirs <- getDirectoryContents oDir
+    externPaths <- traverse (checkExternsPath oDir) dirs
+    pure (catMaybes externPaths)
+
+loadModules
+  :: (Ide m, MonadError PscIdeError m, MonadLogger m)
+  => [P.ModuleName]
+  -> m Success
+loadModules mns = do
+  oDir <- outputDirectory
+  let efPaths = map (\mn -> oDir </> P.runModuleName mn </> "externs.json") mns
+  efiles <- traverse readExternFile efPaths
+  traverse_ insertExterns efiles
+  --TODO Get rid of this once ModuleOld is gone
+  traverse_ insertModule efiles
+  populateStage2
+  pure (TextResult ("Loaded " <> foldMap runModuleNameT mns <> "."))
+
+loadAllModules :: (Ide m, MonadError PscIdeError m) => m Success
 loadAllModules = do
-  outputPath <- confOutputPath . envConfiguration <$> ask
-  cwd <- liftIO getCurrentDirectory
-  let outputDirectory = cwd </> outputPath
-  liftIO (doesDirectoryExist outputDirectory)
-    >>= flip unless (throwError (GeneralError "Couldn't locate your output directory"))
-  liftIO (getDirectoryContents outputDirectory)
-    >>= liftIO . traverse (getExternsPath outputDirectory)
-    >>= traverse_ loadExtern . catMaybes
+  exts <- traverse readExternFile =<< findAllExterns
+  traverse_ insertExterns exts
+  --TODO Get rid of this once ModuleOld is gone
+  traverse_ insertModule exts
+  env <- ask
+  _ <- liftIO $ async (runStdoutLoggingT (runReaderT populateStage2 env))
```

Comment:
Do we need to keep the `Async` around to avoid spawning two stage 2 threads at the same time?